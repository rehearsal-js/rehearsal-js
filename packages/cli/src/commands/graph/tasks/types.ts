export interface GraphWorkerOptions {
  rootPath: string;
  srcPath: string;
  ignore: string[];
  serviceMap: Record<string, string>;
  output?: string;
}

type MessageResponse = { type: 'message'; content: string };
type FilesResponse = { type: 'files'; content: string[] };
export type Response = MessageResponse | FilesResponse;
