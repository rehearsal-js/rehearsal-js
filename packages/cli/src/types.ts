import { MigrationStrategy } from '@rehearsal/migration-graph';

import { UserConfig } from './userConfig';
import { State } from './helpers/state';

export type CliCommand = 'upgrade' | 'migrate';
export type Formats = 'sarif' | 'json' | 'sonarqube' | 'md';

export type MigrateCommandOptions = {
  basePath: string;
  entrypoint: string;
  files: string;
  format: Formats[];
  outputPath: string;
  verbose: boolean | undefined;
  userConfig: string | undefined;
  interactive: boolean | undefined;
};

export type MigrateCommandContext = {
  skip: boolean;
  userConfig: UserConfig | undefined;
  strategy: MigrationStrategy;
  sourceFilesWithAbsolutePath: string[];
  sourceFilesWithRelativePath: string[];
  input: unknown;
  targetPackagePath: string;
  state: State;
};

export type UpgradeCommandContext = {
  tsVersion: string;
  latestAvailableBuild: string;
  currentTSVersion: string;
  skip: boolean;
};

export type UpgradeCommandOptions = {
  build: string;
  basePath: string;
  format: Formats[];
  outputPath: string;
  tsVersion: string;
  dryRun: boolean;
};

export type DependencyConfig = {
  dependencies?: string[];
  devDependencies?: string[];
};

export type SetupConfigCommand = {
  command: string;
  args: string[];
};

export type SetupConfig = {
  ts?: SetupConfigCommand;
  lint?: SetupConfigCommand;
};

export type CustomCommandConfig = {
  install?: DependencyConfig;
  setup?: SetupConfig;
};

export type CustomConfig = {
  upgrade?: CustomCommandConfig;
  migrate?: CustomCommandConfig;
};

export type PackageSelection = {
  name: string;
  path: string;
};

export type MenuMap = {
  [key: string]: string;
};

export type MigrationSummary = {
  totalErrorCount: number;
  hintAddedCount: number;
};
export type TSConfig = {
  compilerOptions: {
    strict: boolean;
  };
};

export type ScriptMap = {
  [key: string]: string;
};
