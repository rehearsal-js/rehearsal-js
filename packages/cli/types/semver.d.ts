declare module 'semver' {
  const semver: {
    valid: (input: string) => boolean;
  };
  export = semver;
}
