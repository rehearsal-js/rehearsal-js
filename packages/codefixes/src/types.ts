import {
  CodeFixAction,
  DiagnosticWithLocation,
  FileTextChanges,
  LanguageService,
  Node,
  Program,
  TypeChecker,
} from 'typescript';

export interface DiagnosticWithContext extends DiagnosticWithLocation {
  service: LanguageService;
  program: Program;
  checker: TypeChecker;
  node?: Node;
}

export interface CodeHintList {
  [key: number]: CodeHintListElement;
}

export interface CodeHintListElement {
  hint: string;
  helpUrl?: string;
  hints?: CodeHint[];
}

export interface CodeHint {
  when: (n: Node, p: Program, c: TypeChecker) => boolean;
  hint: string;
}

export interface CodeFixCollection {
  getFixForDiagnostic(diagnostic: DiagnosticWithContext): CodeFixAction | undefined;
}

export type CodeFixKind = 'insert' | 'delete' | 'replace';

export interface FixedFile {
  fileName: string;
  newCode?: string;
  oldCode?: string;
  codeFixAction?: CodeFixKind;
  location: {
    startLine: number;
    startColumn: number;
    endLine: number;
    endColumn: number;
  };
}

export interface CodeFix {
  getCodeAction: (diagnostic: DiagnosticWithContext) => CodeFixAction | undefined;
}

export function createCodeFixAction(
  fixName: string,
  changes: FileTextChanges[],
  description: string
): CodeFixAction {
  return {
    fixName,
    description,
    changes,
  };
}
