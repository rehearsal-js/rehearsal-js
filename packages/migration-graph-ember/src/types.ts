/* eslint-disable @typescript-eslint/no-explicit-any */
import type { Package } from '@rehearsal/migration-graph-shared';
import type { EmberAddonPackage } from './entities/ember-addon-package.js';
import type { EmberAppPackage } from './entities/ember-app-package.js';

export type EmberProjectPackage = Package | EmberAddonPackage | EmberAppPackage;
