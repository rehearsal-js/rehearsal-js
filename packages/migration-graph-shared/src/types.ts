import type { Package } from './entities/package.js';

export type UniqueNode = {
  key: string;
  synthetic?: boolean;
};

export type PackageNode = UniqueNode & {
  pkg: Package;
  converted?: boolean;
};

export type ModuleNode = UniqueNode & {
  path: string;
  parsed?: boolean;
};
