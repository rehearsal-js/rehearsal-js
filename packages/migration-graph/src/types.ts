import { Package } from '@rehearsal/migration-graph-shared';

import { Graph } from './utils/graph';

export type UniqueGraphNode = {
  key: string;
};

export type PackageNode = UniqueGraphNode & {
  pkg: Package;
  converted?: boolean;
  modules: Graph<ModuleNode>;
};

export type ModuleNode = UniqueGraphNode & {
  path: string;
  parsed?: boolean;
  meta?: unknown;
};
