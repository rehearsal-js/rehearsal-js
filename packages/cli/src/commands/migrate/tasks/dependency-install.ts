import { resolve } from 'node:path';
import { existsSync, promises as fs } from 'node:fs';
import { ListrTask } from 'listr2';

import { addDep } from '@rehearsal/utils';
import { PackageJson } from 'type-fest';
import { MigrateCommandContext, MigrateCommandOptions } from '../../../types.js';

export const REQUIRED_DEPENDENCIES = [
  '@types/node',
  '@typescript-eslint/eslint-plugin',
  '@typescript-eslint/parser',
  'eslint',
  'eslint-config-prettier',
  'eslint-plugin-prettier',
  'prettier',
  'typescript',
];

// get the name of dependency from those two format:
// - foo
// - foo@{version}
// Be aware that a package name would start with @, e.g @types/node
function extractDepName(dep: string): string {
  const reg = /^(@?[^@]+)/g;
  const matched = dep.match(reg);
  return matched ? matched[0] : dep;
}

// check if package.json has all required dependecies
// from rehearsal default and user config
export async function shouldRunDepInstallTask(
  options: MigrateCommandOptions,
  context: Partial<MigrateCommandContext> = {}
): Promise<boolean> {
  const { basePath } = options;
  // add extra required dependencies from user config if there is any
  let dependencies: string[] = [];
  let devDependencies = REQUIRED_DEPENDENCIES;
  if (context.userConfig?.hasDependencies) {
    // check if package.json has all the dependencies listed in user config
    dependencies = [...dependencies, ...context.userConfig.dependencies];
    devDependencies = [...devDependencies, ...context.userConfig.devDependencies];
  }

  const packageJsonPath = resolve(basePath, 'package.json');
  if (existsSync(packageJsonPath)) {
    const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8')) as PackageJson;
    for (const d of dependencies) {
      if (!packageJson.dependencies || !packageJson.dependencies[extractDepName(d)]) {
        return true;
      }
    }

    for (const d of devDependencies) {
      if (!packageJson.devDependencies || !packageJson.devDependencies[extractDepName(d)]) {
        return true;
      }
    }
  } else {
    // no package.json
    return true;
  }
  return false;
}

export function depInstallTask(
  options: MigrateCommandOptions,
  context?: Partial<MigrateCommandContext>
): ListrTask {
  return {
    title: 'Install dependencies',
    enabled: (): boolean => !options.dryRun,
    skip: async (ctx: MigrateCommandContext): Promise<boolean> =>
      !(await shouldRunDepInstallTask(options, ctx)),
    task: async (ctx: MigrateCommandContext, task): Promise<void> => {
      // If context is provide via external parameter, merge with existed
      if (context) {
        ctx = { ...ctx, ...context };
      }

      let dependencies: string[] = [];
      let devDependencies = REQUIRED_DEPENDENCIES;
      // install custom dependencies
      if (ctx.userConfig?.hasDependencies) {
        task.output = `Install dependencies from config`;
        dependencies = [...dependencies, ...ctx.userConfig.dependencies];
        devDependencies = [...devDependencies, ...ctx.userConfig.devDependencies];
        if (ctx.userConfig?.hasPostInstallHook) {
          task.output = `Run postInstall hook from config`;
          await ctx.userConfig.postInstall();
        }
      }

      const errorMessages: string[] = [];

      // for deps
      if (dependencies.length) {
        try {
          task.output = `Installing dependecies: ${dependencies.join(', ')}`;
          await addDep(dependencies, false, { cwd: options.basePath });
        } catch (e) {
          errorMessages.push(formatErrorMessage(dependencies, false));
        }
      }

      // for devDeps
      if (devDependencies.length) {
        try {
          task.output = `Installing devDependencies: ${devDependencies.join(', ')}`;
          await addDep(devDependencies, true, { cwd: options.basePath });
        } catch (e) {
          errorMessages.push(formatErrorMessage(devDependencies, true));
        }

        if (errorMessages.length) {
          throw new Error(errorMessages.join('\n'));
        }
      }
    },
    // will print and keep what dpe is currently installing at bottom bar
    options: { persistentOutput: true, bottomBar: Infinity },
  };
}

function formatErrorMessage(deps: string[], dev: boolean): string {
  const depType = dev ? 'devDependencies' : 'dependencies';
  return [
    `We ran into an error when installing ${depType}, please install the following as ${depType} and try again.`,
    `${deps.map((d) => `  - ${d}`).join('\n')}`,
  ].join('\n');
}
