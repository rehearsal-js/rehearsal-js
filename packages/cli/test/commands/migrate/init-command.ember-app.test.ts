import { resolve } from 'node:path';
import { readdirSync } from 'node:fs';
import { beforeEach, describe, test, expect } from 'vitest';
import { Scenario, PreparedApp } from 'scenario-tester';
import { setGracefulCleanup } from 'tmp';
import { writeJSONSync } from 'fs-extra';
import { appScenarios, clean } from '@rehearsal/test-support';
import { REQUIRED_DEPENDENCIES } from '../../../src/commands/migrate/tasks/dependency-install.js';

import { cleanOutput } from '../../test-helpers/index.js';
import {
  runDefault,
  runWithUserConfig,
  runTwoTimes,
  CUSTOM_CONFIG,
} from '../../test-helpers/init-command-test-utils.js';
import { CustomConfig } from '../../../src/types.js';

setGracefulCleanup();

function createUserConfig(basePath: string, config: CustomConfig): void {
  const configPath = resolve(basePath, 'rehearsal-config.json');
  writeJSONSync(configPath, config);
}

describe('migrate init for ember app variant', () => {
  appScenarios.forEachScenario((scenario: Scenario) => {
    describe(scenario.name, () => {
      let app: PreparedApp;

      beforeEach(async () => {
        app = await scenario.prepare();
        clean(app.dir);
      });

      test(`default run --${scenario.name}`, async () => {
        const {
          stdout,
          devDeps,
          fileList,
          lintConfig,
          lintConfigDefault,
          tsConfig,
          tscLintScript,
        } = await runDefault(app.dir);

        expect(cleanOutput(stdout, app.dir)).toMatchSnapshot();

        for (const devDep of REQUIRED_DEPENDENCIES) {
          expect(Object.keys(devDeps!).includes(devDep));
        }

        expect(fileList).toContain('.eslintrc.js');
        expect(fileList).toContain('.rehearsal-eslintrc.js');

        expect(lintConfig).toMatchSnapshot();
        expect(lintConfigDefault).toMatchSnapshot();
        expect(tsConfig).matchSnapshot();
        expect(tscLintScript).toBe('tsc --noEmit');
      });

      test(`pass user config ${scenario.name}`, async () => {
        createUserConfig(app.dir, CUSTOM_CONFIG);

        const { stdout, devDeps, deps, tscLintScript } = await runWithUserConfig(app.dir);

        expect(cleanOutput(stdout, app.dir)).toMatchSnapshot();

        expect(Object.keys(devDeps!)).includes('tmp');
        expect(deps).toHaveProperty('fs-extra');

        expect(readdirSync(app.dir)).toContain('custom-ts-config-script');
        expect(readdirSync(app.dir)).toContain('custom-lint-config-script');

        expect(tscLintScript).toBe('tsc --noEmit');
      });

      test('skip dep install, ts config, and lint config if exists', async () => {
        const { stdout } = await runTwoTimes(app.dir);

        expect(cleanOutput(stdout, app.dir)).toMatchSnapshot();
      });
    });
  });
});
