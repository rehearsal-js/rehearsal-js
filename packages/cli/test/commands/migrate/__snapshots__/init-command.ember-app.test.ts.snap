// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migrate init for ember app variant > ember4App > default run --ember4App 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > ember4App > default run --ember4App 2`] = `
"module.exports = {
  extends: [\\"./.rehearsal-eslintrc.js\\"],
};
"
`;

exports[`migrate init for ember app variant > ember4App > default run --ember4App 3`] = `
"module.exports = {
  \\"parser\\": \\"@typescript-eslint/parser\\",
  \\"parserOptions\\": {
    \\"sourceType\\": \\"module\\"
  },
  \\"plugins\\": [
    \\"@typescript-eslint\\",
    \\"prettier\\"
  ],
  \\"extends\\": [
    \\"plugin:@typescript-eslint/eslint-recommended\\",
    \\"plugin:@typescript-eslint/recommended\\",
    \\"eslint:recommended\\",
    \\"plugin:prettier/recommended\\",
    \\"prettier\\"
  ]
}"
`;

exports[`migrate init for ember app variant > ember4App > default run --ember4App 4`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "declaration": false,
    "experimentalDecorators": true,
    "importsNotUsedAsValues": "error",
    "inlineSourceMap": true,
    "inlineSources": true,
    "isolatedModules": true,
    "noEmit": true,
    "noImplicitOverride": true,
    "noUncheckedIndexedAccess": true,
    "preserveValueImports": true,
    "strict": true,
    "target": "es2020",
  },
  "display": "Ember App",
  "extends": "../base.tsconfig.json",
  "glint": {
    "environment": {
      "ember-loose": {},
      "ember-template-imports": {},
    },
  },
}
`;

exports[`migrate init for ember app variant > ember4App > pass user config ember4App 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[DATA] Install dependencies from config
[DATA] Installing dependecies: fs-extra
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript, tmp
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[DATA] Create tsconfig from config
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js from config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > ember4App > skip dep install, ts config, and lint config if exists 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[SKIPPED] Install dependencies
[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
[STARTED] Create eslint config
[SKIPPED] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > emberApp > default run --emberApp 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > emberApp > default run --emberApp 2`] = `
"module.exports = {
  extends: [\\"./.rehearsal-eslintrc.js\\"],
};
"
`;

exports[`migrate init for ember app variant > emberApp > default run --emberApp 3`] = `
"module.exports = {
  \\"parser\\": \\"@typescript-eslint/parser\\",
  \\"parserOptions\\": {
    \\"sourceType\\": \\"module\\"
  },
  \\"plugins\\": [
    \\"@typescript-eslint\\",
    \\"prettier\\"
  ],
  \\"extends\\": [
    \\"plugin:@typescript-eslint/eslint-recommended\\",
    \\"plugin:@typescript-eslint/recommended\\",
    \\"eslint:recommended\\",
    \\"plugin:prettier/recommended\\",
    \\"prettier\\"
  ]
}"
`;

exports[`migrate init for ember app variant > emberApp > default run --emberApp 4`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "declaration": false,
    "experimentalDecorators": true,
    "importsNotUsedAsValues": "error",
    "inlineSourceMap": true,
    "inlineSources": true,
    "isolatedModules": true,
    "noEmit": true,
    "noImplicitOverride": true,
    "noUncheckedIndexedAccess": true,
    "preserveValueImports": true,
    "strict": true,
    "target": "es2020",
  },
  "display": "Ember App",
  "extends": "../base.tsconfig.json",
  "glint": {
    "environment": {
      "ember-loose": {},
      "ember-template-imports": {},
    },
  },
}
`;

exports[`migrate init for ember app variant > emberApp > pass user config emberApp 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[DATA] Install dependencies from config
[DATA] Installing dependecies: fs-extra
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript, tmp
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[DATA] Create tsconfig from config
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js from config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > emberApp > skip dep install, ts config, and lint config if exists 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[SKIPPED] Install dependencies
[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
[STARTED] Create eslint config
[SKIPPED] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > emberAppWithInRepoAddon > default run --emberAppWithInRepoAddon 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > emberAppWithInRepoAddon > default run --emberAppWithInRepoAddon 2`] = `
"module.exports = {
  extends: [\\"./.rehearsal-eslintrc.js\\"],
};
"
`;

exports[`migrate init for ember app variant > emberAppWithInRepoAddon > default run --emberAppWithInRepoAddon 3`] = `
"module.exports = {
  \\"parser\\": \\"@typescript-eslint/parser\\",
  \\"parserOptions\\": {
    \\"sourceType\\": \\"module\\"
  },
  \\"plugins\\": [
    \\"@typescript-eslint\\",
    \\"prettier\\"
  ],
  \\"extends\\": [
    \\"plugin:@typescript-eslint/eslint-recommended\\",
    \\"plugin:@typescript-eslint/recommended\\",
    \\"eslint:recommended\\",
    \\"plugin:prettier/recommended\\",
    \\"prettier\\"
  ]
}"
`;

exports[`migrate init for ember app variant > emberAppWithInRepoAddon > default run --emberAppWithInRepoAddon 4`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "declaration": false,
    "experimentalDecorators": true,
    "importsNotUsedAsValues": "error",
    "inlineSourceMap": true,
    "inlineSources": true,
    "isolatedModules": true,
    "noEmit": true,
    "noImplicitOverride": true,
    "noUncheckedIndexedAccess": true,
    "preserveValueImports": true,
    "strict": true,
    "target": "es2020",
  },
  "display": "Ember App",
  "extends": "../base.tsconfig.json",
  "glint": {
    "environment": {
      "ember-loose": {},
      "ember-template-imports": {},
    },
  },
}
`;

exports[`migrate init for ember app variant > emberAppWithInRepoAddon > pass user config emberAppWithInRepoAddon 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[DATA] Install dependencies from config
[DATA] Installing dependecies: fs-extra
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript, tmp
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[DATA] Create tsconfig from config
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js from config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > emberAppWithInRepoAddon > skip dep install, ts config, and lint config if exists 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[SKIPPED] Install dependencies
[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
[STARTED] Create eslint config
[SKIPPED] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > emberAppwithInRepoEngine > default run --emberAppwithInRepoEngine 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js, extending Rehearsal default eslint-related config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > emberAppwithInRepoEngine > default run --emberAppwithInRepoEngine 2`] = `
"module.exports = {
  extends: [\\"./.rehearsal-eslintrc.js\\"],
};
"
`;

exports[`migrate init for ember app variant > emberAppwithInRepoEngine > default run --emberAppwithInRepoEngine 3`] = `
"module.exports = {
  \\"parser\\": \\"@typescript-eslint/parser\\",
  \\"parserOptions\\": {
    \\"sourceType\\": \\"module\\"
  },
  \\"plugins\\": [
    \\"@typescript-eslint\\",
    \\"prettier\\"
  ],
  \\"extends\\": [
    \\"plugin:@typescript-eslint/eslint-recommended\\",
    \\"plugin:@typescript-eslint/recommended\\",
    \\"eslint:recommended\\",
    \\"plugin:prettier/recommended\\",
    \\"prettier\\"
  ]
}"
`;

exports[`migrate init for ember app variant > emberAppwithInRepoEngine > default run --emberAppwithInRepoEngine 4`] = `
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "alwaysStrict": true,
    "declaration": false,
    "experimentalDecorators": true,
    "importsNotUsedAsValues": "error",
    "inlineSourceMap": true,
    "inlineSources": true,
    "isolatedModules": true,
    "noEmit": true,
    "noImplicitOverride": true,
    "noUncheckedIndexedAccess": true,
    "preserveValueImports": true,
    "strict": true,
    "target": "es2020",
  },
  "display": "Ember App",
  "extends": "../base.tsconfig.json",
  "glint": {
    "environment": {
      "ember-loose": {},
      "ember-template-imports": {},
    },
  },
}
`;

exports[`migrate init for ember app variant > emberAppwithInRepoEngine > pass user config emberAppwithInRepoEngine 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[DATA] Install dependencies from config
[DATA] Installing dependecies: fs-extra
[DATA] Installing devDependencies: @types/node, @typescript-eslint/eslint-plugin, @typescript-eslint/parser, eslint, eslint-config-prettier, eslint-plugin-prettier, prettier, typescript, tmp
[SUCCESS] Install dependencies
[STARTED] Create tsconfig.json
[DATA] Create tsconfig from config
[SUCCESS] Create tsconfig.json
[STARTED] Create eslint config
[DATA] Create .eslintrc.js from config
[SUCCESS] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;

exports[`migrate init for ember app variant > emberAppwithInRepoEngine > skip dep install, ts config, and lint config if exists 1`] = `
"[STARTED] Validate project
[SUCCESS] Validate project
[STARTED] Initialize
[DATA] Setting up config for app-template
[SUCCESS] Initialize
[STARTED] Install dependencies
[SKIPPED] Install dependencies
[STARTED] Create tsconfig.json
[DATA] <tmp-path>/tsconfig.json already exists, ensuring strict mode is enabled
[TITLE] Update tsconfig.json
[SUCCESS] Update tsconfig.json
[STARTED] Create eslint config
[SKIPPED] Create eslint config
[STARTED] Add package scripts
[DATA] Adding \\"lint:tsc\\" script in package.json
[SUCCESS] Add package scripts
info:    The project is ready for migration. Please run \\"rehearsal migrate\\" to start the migration."
`;
